src/Compilers/CSharp/Portable/Lowering/Instrumentation/ModuleCancellationInstrumenter.cs

```diff
@@ -0,0 +1,222 @@
+    /// </summary>
+    private MethodSymbol? FindOverloadWithCancellationToken(MethodSymbol method)
+    {
+        // Switching to a cancellable overload only applies to ordinary methods and constructors.
+        if (method.MethodKind is not (MethodKind.Ordinary or MethodKind.Constructor))
+        {
+            return null;
+        }
+
+        // Look for an overload of the definition with no type parameter substitutions.
+        var methodDefinition = method.OriginalDefinition;
+
+        var methodsSetsRequiredMembers = methodDefinition.HasSetsRequiredMembers;
+
+        foreach (var member in methodDefinition.ContainingType.GetMembers(method.Name))
+        {
+            // If the member is the method being compiled, calling it could result in a recursion not present in the code previously.
+            // We can't guarantee the instrumentation won't result in infinite recursion, but we can avoid the trivial case.
+            if (member == _factory.TopLevelMethod?.OriginalDefinition)
+            {
+                continue;
+            }
+
+            // It's unlikely that real-world APIs have overloads that differ in dynamic,
+            // but if they do avoid selecting them since their intended use might differ.
+            //
+            // Similarly, we also only consider overloads with the same visibility.
+            // We could potentially allow the visibility to differ,
+            // but we'd need to check if the overload is accessible at the call site.
+
+            const TypeCompareKind TypeComparisonKind = TypeCompareKind.CLRSignatureCompareOptions & ~TypeCompareKind.IgnoreDynamic;
+
+            if (member.IsStatic == methodDefinition.IsStatic &&
+                member.MetadataVisibility == methodDefinition.MetadataVisibility &&
+                member is MethodSymbol { Parameters: [.., { RefKind: RefKind.None, Type: { } lastParamType }] parametersWithCancellationToken } overload &&
+                overload.Arity == methodDefinition.Arity &&
+                methodDefinition.MethodKind == overload.MethodKind &&
+                methodDefinition.Parameters.Length == parametersWithCancellationToken.Length - 1 &&
+                lastParamType.Equals(_throwMethod.ContainingType, TypeCompareKind.ConsiderEverything) &&
+                (!methodsSetsRequiredMembers || overload.HasSetsRequiredMembers))
+            {
+                var typeMap = (methodDefinition.Arity > 0) ? new TypeMap(overload.TypeParameters, methodDefinition.TypeParameters) : null;
+
+                if (MemberSignatureComparer.HaveSameParameterTypes(
+                        methodDefinition.Parameters.AsSpan(),
+                        typeMap1: null,
+                        parametersWithCancellationToken.AsSpan(0, methodDefinition.Parameters.Length),
+                        typeMap,
+                        MemberSignatureComparer.RefKindCompareMode.ConsiderDifferences,
```

Comment:
I am wondering if any unexpected behavior changes could occur if:
- the original method was a `readonly` struct method and the replacement method was not, for example.
- Or, the other way around, the original is non-readonly and the replacement is readonly.

Since it seems like we are not checking the readonly-ness of the `this` parameter in these checks.

Since this PR is already merged some time ago, I will open a new issue based on this comment.